corehq/motech/value_source.py

```diff
@@ -451,6 +449,124 @@ class CasePropertyConstantValue(ConstantValue, CaseProperty):
     pass
 
 
+@attr.s(auto_attribs=True, kw_only=True)
+class SupercaseValueSource(ValueSource):
+    """
+    A reference to a list of parent/host cases.
+
+    Evaluates nested ValueSource config, allowing for recursion.
+    """
+    supercase_value_source: dict
+
+    # Optional filters for indices
+    identifier: Optional[str] = None
```

Comment:
What is the semantic difference between this and `identifier: str = None`? Is it just wrong to set the default to something that violates the type annotation?

Also, how is a type annotation like this practically useful other than strictly as documentation? Does your IDE give you rich auto-complete information for the union of `None` and `str`?

I promise these are genuine questions, and I'm not trying to fortify a case against type annotations.

Reply:
> What is the semantic difference between this and `identifier: str = None`? Is it just wrong ...

Yes. A bit rude telling you it's a string, and then setting it to a value that isn't a string. But there's more to it than that ...

> Also, how is a type annotation like this practically useful other than strictly as documentation?

... `@attr.s(auto_attribs=True)` (above) tells **attrs** to use type annotations to define attributes instead of calls to `attr.ib()`. So this does more than just serve as documentation. (You can find more info at the [**attrs** docs](https://www.attrs.org/en/stable/types.html).)

Reply:
Ahh, I overlooked the `@attr.s(auto_attribs=True)` decorator. That makes sense. Thanks for the explanation.